### PR TITLE
:sparkles: migrate to k-ramel

### LINKED ISSUE #104: migrate redux / saga to k-ramel

### PR DESCRIPTION
**Pull request by bpetetot**
Closes #104

### CODE REVIEW

**Comment by fabienjuif** [comment]
How do you feel this migration right now ?

**Comment by bpetetot** [comment]
I am doing the migration in several steps : 
- [X] 1. apply `createStore` and `inject` of k-ramel (without refactoring) (836087dbb3b04865e4dab32223a84d1e5452d7a5)
   _works directly with `hideRedux=false`_ :heart:
- [x] 2. apply `hideRedux=true` and make it working (38818dfa33d5635923a1ad90b278a4dcfeba6197)
- [x] 3. replace redux reducers by `k-redux-factory`
- [x] 4. replace saga by listeners progressively and refactor actions names (it's messy today)
- [x] 5. apply listeners closer to the screens

File: src/screens/organizer/proposals/proposalFilters/proposalFilters.container.js [hunk 1]
```diff
@@ -1,19 +1,20 @@
-import { connect } from 'react-redux'
+import { inject } from 'k-ramel/react'
```
Comment (fabienjuif) [inline src/screens/organizer/proposals/proposalFilters/proposalFilters.container.js]: I am thinking about publishing x distincts packages:
 - `k-ramel`
 - `@k-ramel/react`
 - `@k-ramel/driver-http`
 - `@k-ramel/...`

What is your though ?

Comment (bpetetot) [reply src/screens/organizer/proposals/proposalFilters/proposalFilters.container.js]: Yes it's a good idea, but I think you should keep an unique github repository with sub-packages

Comment (fabienjuif) [reply src/screens/organizer/proposals/proposalFilters/proposalFilters.container.js]: Yes that is the idea :)

File: src/screens/speaker/myTalks/myTalks.container.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ import MyTalks from './myTalks'
 const mapStore = store => ({
   loaded: store.ui.speaker.myTalks.isInitialized(),
   talks: store.ui.speaker.myTalks.getAsArray(),
-  load: () => store.dispatch({ type: 'ON_LOAD_SPEAKER_TALKS' }),
+  load: () => store.dispatch({ type: '@@ui/ON_LOAD_SPEAKER_TALKS' }),
```
Comment (fabienjuif) [inline src/screens/speaker/myTalks/myTalks.container.js]: You can `dispatch('@@ui/ON_LOAD_SPEAKER_TALKS')` now if you want :)

Comment (bpetetot) [reply src/screens/speaker/myTalks/myTalks.container.js]: OMG ! I forgot that. its fantastic !

**Comment by bpetetot** [comment]
@fabienjuif @guillaumecrespel

`k-ramel` migration finished and `saga` out !
About 1/3 of code deleted

Now I must do some tests to see if I have some regression, but the migration was not difficult and progressive.

**Comment by guillaumecrespel** [comment]
is it possible to use your example in our talk ?
image

**Comment by fabienjuif** [comment]
There still have some refactoring to do @bpetetot
I think about drivers.
You use a lot of dispatch in your reactions.

We have a custom driver (that we have to open source) for redux-form, it is like this:
```js
import { setSubmitSucceeded, getFormValues, setSubmitFailed, startSubmit, stopSubmit, isSubmitting } from 'redux-form'

const asyncSubmit = (name, dispatch, getState) => async (http, ...options) => {
  dispatch(startSubmit(name))
  await http(...options)
  if (isSubmitting(name)(getState())) dispatch(stopSubmit)
}

export default ({ dispatch, getState }) => name => ({
  getFormValues: () => getFormValues(name, state => state.ui.form)(getState()),
  setSubmitFailed: (...fields) => dispatch(setSubmitFailed(name, ...fields)),
  setSubmitSucceeded: () => dispatch(setSubmitSucceeded(name)),
  startSubmit: () => dispatch(startSubmit(name)),
  stopSubmit: errors => dispatch(stopSubmit(name, errors)),
  asyncSubmit: (http, ...options) => asyncSubmit(name, dispatch, getState)(http, ...options),
})
```

And can be use like this for example:
```js

export const login = reaction((action, store, { http, form }) => {
  // retrieve credentials from redux
  const signin = form('signin')
  const values = signin.getFormValues()
  // call API
  signin.asyncSubmit(
    http('AUTH')
    '/api/login',
    {
      method: 'POST',
      body: JSON.stringify(values),
      headers: { 'Content-Type': 'application/json' },
    },
  )
})
```

**Comment by bpetetot** [comment]
**Yeah ! that's great ! and yes I have a lot of refactoring to do :** 

- [x] use redux-form driver ❤️ 

- [ ] find a way to manage data loading on routes and component mounting. Each time I have an URL with param `eventId`, I have to do a loader like that, I think I could factorize this code in a better way.
```es6
const mapStore = (store) => {
  const eventId = getRouterParam('eventId')(store.getState())
  const event = store.data.events.get(eventId)
  return {
    loaded: !!event,
    eventId,
    load: () => store.dispatch('@@ui/ON_LOAD_EVENT'),
  }
}
export default compose(
  forRoute.absolute('PROPOSALS'),
  inject(mapStore),
  loader,
)(Proposals)
```

- [ ] maybe create a driver to have a simpliest use of the router in reaction : 
```es6
 // replace
 const eventId = getRouterParam('eventId')(store)
 // by
 const eventId = router.getParam('eventId')

 // replace
 store.dispatch(push(`/organizer/event/${eventId}`))
 // by
 router.push(`/organizer/event/${eventId}`)
```

- [ ] rename some action names, it's still messy in my mind sometimes

**Comment by fabienjuif** [comment]
For the second point, look at the new hoc `listen`.
Maybe we can dispatch an event in this hoc like `@@krml/LISTEN_PLUGUED>${componentName}`

And in the listeners you plug with `listen` hoc, you catch that ?

**Comment by bpetetot** [comment]
@guillaumecrespel yes you can ;)

File: src/screens/speaker/talk/edit/talkEdit.container.js [hunk 3]
```diff
@@ -6,17 +6,15 @@ import { getRouterParam } from 'store/reducers/router'
...
 
-const FORM_NAME = 'talk-edit'
-
 const mapStore = (store) => {
   const talkId = getRouterParam('talkId')(store.getState())
   const talk = store.data.talks.get(talkId)
   return {
     loaded: !!talk,
-    form: FORM_NAME,
+    form: 'talk-edit',
```
Comment (fabienjuif) [inline src/screens/speaker/talk/edit/talkEdit.container.js]: You are throwing away the optimisation with a string here I think.
This was better to make it a constant so the string reference is always the same and the shallowCompare works better

File: src/store/drivers/reduxForm.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import {
```
Comment (fabienjuif) [inline src/store/drivers/reduxForm.js:1]: You should wait k-ramel driver before merging IMO

File: src/store/listeners/reactions/events.js [hunk 5]
```diff
@@ -1,48 +1,28 @@
...
-    store.dispatch(reset(FORM))
-    // set form submitted
-    store.dispatch(stopSubmit(FORM))
-    // go to event page
-    store.dispatch(push(`/organizer/event/${ref.id}`))
-  } catch (error) {
-    store.dispatch(stopSubmit(FORM, { _error: error.message }))
-    throw error
-  }
+export const createEvent = reaction(async (action, store, { form }) => {
```
Comment (fabienjuif) [inline src/store/listeners/reactions/events.js]: This is better !

File: src/store/listeners/reactions/events.js [hunk 6]
```diff
@@ -1,48 +1,28 @@
...
-      throw error
-    }
+export const updateEvent = formName =>
+  reaction((action, store, { form }) => {
+    const updateForm = form(formName)
+    const event = updateForm.getFormValues()
+    // update event into database
+    updateForm.asyncSubmit(eventCrud.update, event)
+    // update event in store
+    store.data.events.update(event)
```
Comment (fabienjuif) [inline src/store/listeners/reactions/events.js:25]: Don't you want to await asyncSubmit ?

File: src/store/listeners/reactions/talks.js [hunk 7]
```diff
@@ -1,51 +1,30 @@
 import { reaction } from 'k-ramel'
```
Comment (fabienjuif) [inline src/store/listeners/reactions/talks.js]: I think we agree, but this file should be attach with listen hoc no ?
And others of same types

File: src/store/drivers/reduxForm.js (hunk 4 above)
Comment (bpetetot) [reply src/store/drivers/reduxForm.js:1]: Yes I will wait for it ;)

File: src/store/listeners/reactions/events.js (hunk 6 above)
Comment (bpetetot) [reply src/store/listeners/reactions/events.js:25]: No, I'm an optimistic guy 😃

File: src/store/listeners/reactions/talks.js (hunk 7 above)
Comment (bpetetot) [reply src/store/listeners/reactions/talks.js]: Yes it will be an other refactoring step...

**Comment by fabienjuif** [comment]
❤️

**Pull request merged**